### PR TITLE
Add static deployment route matching

### PR DESCRIPTION
**Pull request by kiootic**
ref #1201

### CODE REVIEW

File: pkg/gateway/model/route_match_test.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		}
+		So(toURL("/", "http://backend"), ShouldEqual, "http://backend/")
+		So(toURL("/", "http://backend/"), ShouldEqual, "http://backend/")
+		So(toURL("/api", "http://backend"), ShouldEqual, "http://backend/api")
+		So(toURL("/api/", "http://backend"), ShouldEqual, "http://backend/api/")
+		So(toURL("/", "http://backend/api"), ShouldEqual, "http://backend/api")
+		So(toURL("/", "http://backend/api/"), ShouldEqual, "http://backend/api/")
+		So(toURL("/login", "http://backend/api"), ShouldEqual, "http://backend/api/login")
+		So(toURL("/login/", "http://backend/api"), ShouldEqual, "http://backend/api/login/")
+		So(toURL("/login", "http://backend/api"), ShouldEqual, "http://backend/api/login")
```
Comment (louischan-oursky) [inline pkg/gateway/model/route_match_test.go]: This line is the same as line 92.

File: pkg/gateway/handler/deployment_route_test.go [hunk 2]
```diff
@@ -10,157 +11,63 @@ import (
...
-		config.DeploymentRoute{
-			Path: "/api/",
-			Type: model.DeploymentRouteTypeHTTPService,
-			TypeConfig: map[string]interface{}{
-				"backend_url": "http://backend-domain/",
+				url:       "https://example.com/api/login?user=test&password=1234",
+				matchPath: "/login",
+				route: config.DeploymentRoute{
+					Type: "http-service",
+					TypeConfig: map[string]interface{}{
```
Comment (louischan-oursky) [inline pkg/gateway/handler/deployment_route_test.go]: Some DeploymentRoute fixtures are missing `Path`. Is it intended?

File: pkg/gateway/model/deployment_route.go [hunk 3]
```diff
@@ -32,3 +33,19 @@ func (r RouteTypeConfig) TargetPath() string {
 	}
 	return ""
 }
+
+func (r RouteTypeConfig) AssetPathMapping() map[string]string {
+	m := r["asset_path_mapping"].(map[string]interface{})
```
Comment (carmenlau) [inline pkg/gateway/model/deployment_route.go]: Can you add some explanation how we construct `asset_path_mapping`?

Comment (kiootic) [reply pkg/gateway/model/deployment_route.go]: You can see the `deployment_route_test.go` file for reference on how `static` deployment route works. Copied below for reference:
```go
{
	url:       "https://example.com/login#form",
	matchPath: "/index-html",
	route: config.DeploymentRoute{
		Type: "static",
		Path: "/",
		TypeConfig: map[string]interface{}{
			"backend_url": "https://app.localhost/_asset/get",
			"asset_path_mapping": map[string]interface{}{
				"/index.html":              "index-html",
				"/assets/main.12345678.js": "main-js",
			},
			"asset_fallback_path": "/",
		},
	},
	forwardURL: "https://app.localhost/_asset/get/index-html",
}
```

`asset_fallback_path` corresponds to `fallback` config in `skygear.yaml`.
`asset_path_mapping` is a map with URL absolute path (based on deployment route path) as key, and asset name as value.

Controller would fill in the get asset endpoint of app asset gear as `backend_url`, and upload individual static assets to it.

File: pkg/gateway/model/route_match.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		attempt++
+
+		match := matchRoutePath(reqPath, routes)
+		if match == nil {
+			return nil
+		}
+		if match.Route.Type == DeploymentRouteTypeStatic {
+			config := RouteTypeConfig(match.Route.TypeConfig)
+			pathMapping := config.AssetPathMapping()
+			assetPath := path.Clean("/" + reqPath)
```
Comment (carmenlau) [inline pkg/gateway/model/route_match.go]: Should it be `match.Path` instead of `reqPath` here?

Comment (carmenlau) [reply pkg/gateway/model/route_match.go]: Or the `asset_path_mapping` keys will have the path prefix?

Comment (kiootic) [reply pkg/gateway/model/route_match.go]: Fixed. Should be using `match.Path`.

**Pull request merged**